숫자게임 [Step2] sookoong, groot

Jeon-Minsu opened the pull request:
[@dnwhd0112](https://github.com/dnwhd0112)
안녕하세요 우롱차!! Step2에 대한 PR을 보내는 groot와 수꿍입니다! 바쁘심에도 불과하고 열심히 리뷰해주셔서 감사합니다!  저번 리뷰도 정말 많은 도움이 되었습니다! 이번 리뷰도 잘 부탁드립니다!! 😀😀

## [고민한 부분]

  inputUserNumber 함수를 만들면서 guard let을 여러번 이어 붙여서 값에 대한 검증을 진행하려고 했습니다.

(현재 작성한 코드)

```swift

while inputArray.count < 3 {
        print("숫자 3개를 띄어쓰기로 구분하여 입력해주세요.", "중복 숫자는 허용하지 않습니다.", "입력 : ", separator: "\n", terminator: "")
        
        userNumber = readLine()?.split(separator: " ").map{Int($0) ?? 0}
        
        guard let unwrappedUserNumber: Array<Int> = userNumber?.filter({$0 >= 1 && $0 <= 9}) else {
            return
        }
        
        if Set(unwrappedUserNumber).count >= 3 {
            inputArray = unwrappedUserNumber
        } else {
            print("입력이 잘못되었습니다")
        }

```

(일전에 작성하다 에러가 난 코드)
```swift
func inputUserNumber() {
    let userNumber = readLine()?.split(separator: " ").map{Int($0) ?? 0}
    guard let unwrappedUserNumber = userNumber else {
        print("숫자 3개를 띄어쓰기로 구분하여 입력해주세요.", "중복 숫자는 허용하지 않습니다.")
        return
    }
    guard Set(unwrappedUserNumber.count) > 2 else {
        print("숫자 3개를 띄어쓰기로 구분하여 입력해주세요.", "중복 숫자는 허용하지 않습니다.")
        return
    }
    for number in 0...2
    {
        guard unwrappedUserNumber[number] >= 1, unwrappedUserNumber[number] <= 9 else {
            print("숫자 3개를 띄어쓰기로 구분하여 입력해주세요.", "중복 숫자는 허용하지 않습니다.")
            return
        }
    }
    inputArray = unwrappedUserNumber
}
```

  먼저, 에러가 난 코드는 반복문을 쓰지 않았지만 이를 무시하고 핵심 기능에 집중하여 설명드리자면 다른 부분 코드가 완벽하지 않은 상태에서 guard문을 사용한 코드를 돌리다 보니 에러가 났습니다. 이에 저희는 에러가 난 다른 코드가 아닌 guard문 내에서 else 값이 예외 처리되지 않고 그대로 통과하는 문제가 있다고 잘못 판단하고 말았습니다. 저희가 guard문에 대하여 착실히 공부하여 확신이 있었다면 다른 쪽에 문제가 있는지 시각을 넓힐 수 있었으나 이에 매몰되고 말았습니다. 후에 코드를 위를 기반으로 새로 짜 본 결과 정상적으로 구동함을 확인하였습니다. 이 경험을 바탕으로 guard문을 더욱 공부해야 함을 느꼈습니다.


## [궁금한 부분]

1. guard문 내 return의 의미

  위의 고민한 부분의 연장선으로 저희는 guard문 내 return의 의미를 분석하고 있는 가운데, 종료 지시어로 사용되는 것 이외에는 그 의미를 찾는데 어려움을 겪고 있습니다. 이에 종료 지시어 이외에 다른 의미가 있는지 여쭈어 봅니다.

2. 출력 함수 코드 분할 고민

  사용자 입력 메뉴를 실행하는 selectMenu 함수 중, 74번째 줄과 관련하여
출력함수를 전자와 같이 한번에 쓰는 것이 깔끔할 지, 후자와 같이 나누는 것이 더욱 가독성이 있는지 질문드립니다.

(전자)
```swift
print("1. 게임시작", "2. 게임종료", "원하는 기능을 선택해주세요 : ",  separator: "\n", terminator: "")
```

(후자)
```swift
print("1. 게임시작", "2. 게임종료", separator: "\n",)
print(“원하는 기능을 선택해주세요 : ",   terminator: “”)
```

3. 이중 종료 문제 고민

  사용자 입력 메뉴를 실행하는 selectMenu 함수 중, repeat ~ while 문 중 89번째 줄에서 loop가 종료하게 다음과 같이 while문을 구성하였습니다. 

```swift
while mainMenu != "2" 
```

그러나 81번째 줄의 break 명령어를 통해서 반복을 종료하는데 이를 그냥 빈칸으로 두어야 하는지, 아니면 더 좋은 대체 방안이 있는지 궁금하여 질문드립니다.

Jeon-Minsu committed:
feat: 사용자 메뉴를 입력받고 메뉴를 출력하는 함수 생성

- selectMenu 함수를 생성하여 사용자 메뉴를 입력받고 출력
- switch-case문을 이용하여 게임시작과 게임종료, 입력을 잘못받은 경우를 처리

Groot-94 committed:
feat: inputUserNumber 함수에 사용자의 입력에 대한 유효성을 검증하는 기능추가

Jeon-Minsu committed:
refactor: 임의의 수 생성에서 사용자 입력으로 변경 및 게임 결과 호출 출력 순서 수정

- compareNumber 함수 내에서 inputArray에 generateRandomNumber 함수를 받지 않고 inputuserNumber 함수를 이용하여 사용자 입력을 받도록 수정
- 사용자와 컴퓨터 승리 판단 후 스트라이크, 볼, 남은 기회를 출력하는 방식에서 요구사항에 따라 스트라이크, 볼을 먼저 출력하고 이후에 조건에 따라 사용자 혹은 컴퓨터 승리, 남은 기회를 출력하는 방식으로 수정
- 더 나은 직관성을 위하여 기타 코드 자체 및 간격 수정

Groot-94 committed:
docs: Step2 순서도 작성

Jeon-Minsu committed:
docs: README 파일에 STEP2 순서도 추가

dnwhd0112 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -23,30 +23,45 @@ func generateRandomNumber() -> Array<Int> {
+    inputArray = []
+    
+    while inputArray.count < 3 {
+        print("숫자 3개를 띄어쓰기로 구분하여 입력해주세요.", "중복 숫자는 허용하지 않습니다.", "입력 : ", separator: "\n", terminator: "")
```
여러줄의 print 문을 다른방법으로 쓰는건 어떨까요?

Groot-94 replied:
""" 를 이용해 수정해봤습니다 :) 👍

dnwhd0112 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -23,30 +23,45 @@ func generateRandomNumber() -> Array<Int> {
+            return
+        }
+        
+        if Set(unwrappedUserNumber).count >= 3 {
```
여기서 set으로 변환하는 이유가 있을까요???

Groot-94 replied:
입력받는 숫자의 배열에서 중복된 숫자를 Set을 통해 제거하고, 제거했을 때 원소의 갯수가 3이거나 크다면 inputArray에 넣어주는걸 표현하기 위해서 사용했습니다.
set으로 변환하지 않으면 중복된 숫자들의 입력을 체크할 수 없어서 저렇게 작성했습니다.

dnwhd0112 replied:
왜 set으로 변환하지 않으면 중복된 숫자들의 입력을 체크할수가 없나요??? 혹시 array의 contains 함수를 찾아보셨나요?

dnwhd0112 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -23,30 +23,45 @@ func generateRandomNumber() -> Array<Int> {
+        
+        userNumber = readLine()?.split(separator: " ").map{Int($0) ?? 0}
+        
+        guard let unwrappedUserNumber: Array<Int> = userNumber?.filter({$0 >= 1 && $0 <= 9}) else {
```
고차함수의 활용 좋습니다!
여기서 타입을 써주신 이유가 있을까요?

Jeon-Minsu replied:
각 변수가 어떤 자료형인지 명확히 표현하기위하여 타입을 작성해보았습니다!

dnwhd0112 replied:
명확히 표현한다는것은 혼동의 여지가 있기때문이라고 생각하신거죠? 그거 때문에 쓰셨다면 이해가 갑니다.

dnwhd0112 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -23,30 +23,45 @@ func generateRandomNumber() -> Array<Int> {
+    while inputArray.count < 3 {
+        print("숫자 3개를 띄어쓰기로 구분하여 입력해주세요.", "중복 숫자는 허용하지 않습니다.", "입력 : ", separator: "\n", terminator: "")
+        
+        userNumber = readLine()?.split(separator: " ").map{Int($0) ?? 0}
```
여기서 숫자가 아닌 문자를 입력하거나 readLine이 nil을 반환하면 어떻게 되나요?

Groot-94 replied:
Nil-Coalescing Operator 을 사용하여 옵셔널 추출을 하기 때문에 숫자가 아닌 문자를 입력받거나, nil을 반환하면 기본값으로 설정된  0이 들어오게 됩니다.

dnwhd0112 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -23,30 +23,45 @@ func generateRandomNumber() -> Array<Int> {
         
+        inputUserNumber()
         checkStrike()
         checkBall()
```
함수로 대체한 점 아주 좋습니다.

Jeon-Minsu replied:
감사합니다!

dnwhd0112 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -59,12 +74,31 @@ func checkStrike() {
+        
+        mainMenu = readLine() ?? ""
+        
+        switch mainMenu {
```
switch 문으로 입력을 처리한점 좋습니다.

Jeon-Minsu replied:
좋은 방법으로 처리해서 다행이네요!

dnwhd0112 reviewed (changes requested):
안녕하세요 수꿍, 그루트

코드를 보다보니 매직 넘버 매직 스트링을 사용하는 방법도 좋게 느껴져요! 한번 찾아보시고 적용해보시면 좋겠어요.

1. guard문 내 return의 의미
guard문은 이 조건을 만족하지 않으면 더 이상 진행하지 않겠어! 하는 느낌이 있습니다.
return문을 안쓰면 뜨는 에러 경고문도 한번 보시고 밑에있는 문서도 한번 보시면 좋을꺼 같습니다.
(https://docs.swift.org/swift-book/LanguageGuide/ControlFlow.html)

2. 출력 함수 코드 분할 고민
제가 생각한건 ```을 사용한 방법이였는대 여러개의 String표현법을 생각해보시고 더 좋다고 느끼시는걸 채택해보세요! 정답은 없습니다.

3. break 대신에 종료 안내문구를 쓰는건 어떨까요?
while문에서도 2가 나오면 종료가 되는대 break 도 쓰는것은 이중으로 처리하는 것처럼 보이네요.

Groot-94 committed:
refactor: 매직넘버를 심볼릭 정수로 변환

- requiredNumberOfElements 게임에 사용할 숫자의 개수 설정
- minimumNumber 게임에서 사용할 숫자의 최솟값을 설정
- maximumNumber 게임에서 사용할 숫자의 최댓값을 설정
- goalOfStrikeCount 목표로 설정할 스트라이크의 개수 설정
- gameOverCount 남은 기회를 모두 소진했을 경우의 수 설정

Jeon-Minsu committed:
refactor: 매직 스트링을 심볼릭 스트링으로 변환, 변수명 수정, 함수 순서 변경

- 매직 스트링 사용을 피하기 위하여 "입력이 잘못되었습니다"를 inputError라는 상수에 할당
- selectMenu함수를 executeNumberBaseballGame으로 수정
- mainMenu 변수명을 userInput으로 수정
- 한줄로 입력된 게임시작, 게임종료, 원하는 기능 선택 창을 """을 이용하여 여러 줄로 한번에 표현 및, terminator를 ""에서 " "로 수정
- userInput의 값을 판단하는 switch문을 performUserSelect함수로 빼내어 두 번 들여쓰기를 지양
- performUserSelect 함수 내 case "2"의 경우 break로 인하여 이중 종료되는 것을 종료 문구 사용으로 대체
- generateRandomNumber 함수 내 while 조건문의 부등호 방향이 '>='로 잘못되어 '<='로 수정

- compareNumber 함수 내 사용자승리, 컴퓨터승리, 남은기회를 판단한는 조건문을 printResult함수로 빼내어 두 번 들여쓰기를 지양
- inputUserNumber 함수 내 while 문 내 유효성 검사 출력 문구를 """를 사용하여 여러 줄로 한번에 표현 및, terminator를 ""에서 " "로 수
- compareNumber 함수 내 striekCount를 0으로 초기화하는 코드를 checkStrike 함수 내로 위치 변경
- compareNumber 함수 내 스트라이크와 볼을 출력하는 코드를 checkBall 함수 내로 위치 변경
- compareNumber 함수 내 remainingChances를 하나씩 차감한 코드를 printReuslt 함수 내로 위치 변경

Groot-94 committed:
refactor: 함수기능 분리, 변수 설정부분 위치변경

- 사용자의 입력을 검증하는 부분을 checkValidity 함수로 기능분리
- inputUserNumber 함수에서 inputArray = [] 제거
- checkBall 함수에 inputArray.removeAll() 추가
- checkBall 함수에 remainingChances -= 1 추가
- strike와 ball의 결과를 출력하는 부분을 printResult 함수로 이동

dnwhd0112 commented:
매직스트링을 사용해서 대체하신점 좋습니다. 다만 나중에 수정을 할때 따로 모아놓은 파일에서 보면 더 수정하기 쉽지 않을까요?
저는 보통 enum파일을 따로 만들어서 관리하는 편이에요! 이 부분은 찾아보시고 적용하시면 좋을꺼 같습니다.
